src/smartstream/derive/src/ast.rs

```diff
@@ -0,0 +1,61 @@
+use syn::{AttributeArgs, NestedMeta, Meta, ItemFn, Error as SynError, Result as SynResult};
+use syn::spanned::Spanned;
+use proc_macro2::Ident;
+
+/// The configuration for the Smart Stream that will be generated.
+///
+/// This is constructed from the AttributeArgs of the derive macro.
+#[derive(Debug)]
+pub struct SmartStreamConfig {
+    pub kind: SmartStreamKind,
+}
+
+impl SmartStreamConfig {
+    pub fn from_ast(args: AttributeArgs) -> SynResult<Self> {
+        let kind = SmartStreamKind::from_ast(args)?;
+        Ok(Self { kind })
+    }
+}
+
+#[derive(Debug)]
+pub enum SmartStreamKind {
+    Filter,
+    Map,
+    FlatMap,
+}
+
+impl SmartStreamKind {
+    fn from_ast(args: AttributeArgs) -> SynResult<Self> {
+        let ss_type =
+            args.iter()
+                .filter_map(|it| match it {
+                    NestedMeta::Meta(Meta::Path(it)) => {
+                        it.segments.iter().rev().next().and_then(|it| {
+                            match &*it.ident.to_string() {
+                                "filter" => Some(SmartStreamKind::Filter),
+                                "map" => Some(SmartStreamKind::Map),
+                                "flat_map" => Some(SmartStreamKind::FlatMap),
+                                _ => None,
+                            }
+                        })
+                    }
+                    _ => None,
+                })
+                .next()
+                .ok_or_else(|| SynError::new(args[0].span(), "Missing SmartStream type"))?;
+
+        Ok(ss_type)
+    }
+}
+
+pub struct SmartStreamFn {
+    pub name: Ident,
+    pub func: ItemFn,
+}
+
+impl SmartStreamFn {
+    pub fn from_ast(func: ItemFn) -> SynResult<Self> {
+        let name = func.sig.ident.clone();
```

Comment:
no need for clone here

Reply:
Without the clone this does not compile, because I am storing both func and name in a struct, and taking the ident from the func.sig is a partial move

Reply:
fair enough